src/amoc_config.erl

```diff
@@ -5,41 +5,103 @@
 
 %% ------------------------------------------------------------------
 %% API
 %% ------------------------------------------------------------------
--spec get(atom()) -> any().
+-spec get(name()) -> any().
 get(Name) ->
     get(Name, undefined).
 
--spec get(atom(), any()) -> any().
+-spec get(name(), any()) -> any().
 get(Name, Default) ->
     case os:getenv("AMOC_" ++ erlang:atom_to_list(Name)) of
         false ->
             application:get_env(amoc, Name, Default);
         Value ->
             parse_value(Value)
-    end.       
+    end.
 
--spec fetch(atom()) -> any().
-fetch(Name) ->
-    case os:getenv("AMOC_" ++ erlang:atom_to_list(Name)) of
-        false ->
-            case application:get_env(amoc, Name, undefined) of
-                undefined ->
-                    throw({key_not_found, Name});
-                Value ->
-                    Value
-            end;
-        Value ->
-            parse_value(Value)
+-spec parse_scenario_settings(scenario_configuration()) -> {ok, settings()} | {error, invalid_settings()}.
+parse_scenario_settings(Config) ->
+    ParametersVerification = [get_value_and_verify(C) || C <- Config],
+    case [{N, V, R} || {N, V, {false, R}} <- ParametersVerification] of
+        [] ->
+            KeyValList = [{Name, Value} || {Name, Value, _} <- ParametersVerification],
+            {ok, maps:from_list(KeyValList)};
+        InvalidSettings -> {error, {invalid_settings, InvalidSettings}}
+    end.
+
+-spec get_scenario_parameter(name(), settings()) -> {ok, value()} | {error, no_parameter}.
+get_scenario_parameter(Name, Settings) ->
+    case maps:find(Name, Settings) of
+        {ok, Value} -> {ok, Value};
+        error -> {error, no_parameter}
     end.
+
 %% ------------------------------------------------------------------
 %% Internal Function Definitions
 %% ------------------------------------------------------------------
+get_value_and_verify({Name, Env, DefaultValue, VerificationMethod}) ->
+    Value = amoc_config:get(Env, DefaultValue),
+    DefaultValueVerification = verify_value(DefaultValue, VerificationMethod),
+    ValueVerification = verify_value(Value, VerificationMethod),
+    Verification = case {DefaultValueVerification, ValueVerification} of
+                       {true, true} -> true;
+                       {false, true} ->
+                           lager:error("Invalid default value for ~p", [Name]),
+                           {false, bad_default_value};
+                       {true, false} ->
+                           lager:error("Invalid value for ~p", [Name]),
+                           {false, bad_value};
+                       {false, false} ->
+                           lager:error("Invalid default value & value for ~p", [Name]),
+                           {false, bad_value_bad_default_value}
+                   end,
+    {Name, Value, Verification}.
+
+verify_value(_, none) -> true;
```

Comment:
lets create amoc_config_validation module with all these:
* none/1
* positive_integer/1
* nonnegative_integer/1
* bitstring/1

and have a general implementation like this one:
```

verify_value(Value, Atom) when is_atom(Atom) ->
    Fun = fun(Value) ->
            erlang:apply(amoc_config_validation, Atom, [Value])
          end,
    call_verify_fun(Fun, Value)
verify_value(Value, [_ | _] = NonemptyList) ->
    is_one_of(Value, NonemptyList);
verify_value(Value, Fun) when is_function(Fun, 1) ->
    call_verify_fun(Fun, Value);
verify_value(_, VerificationMethod) ->
    lager:error("invalid verification method ~p", [VerificationMethod]),
    false.

call_verify_fun(Fun, Value) ->
    try Fun(Value) of
        Bool when is_boolean(Bool) -> Bool;
        _ ->
            lager:error("invalid verification method ~p", [Fun]),
            false
    catch
        C:E:S ->
            lager:error("invalid verification method ~p ~p ~p", [C,E,S]),
            false
    end

```

so it will be easy to extend the list of validation functions. MB even make the module name configurable via erlang application env.

Reply:
I think we are already flexible enough here because of the `fun()` option. If needed another validation module could be passed as a scenario parameter. The validator can specified like:
* `fun erlang:is_bitstring/1` or
* `fun(V) -> erlang:apply(amoc_config:get_scenario_param(validation_mod), validatation_func, [Value]) end`

I know it may not be very convenient but I would wait with such a change until we see how this validation is used in practice. For now we have only two very similar scenarios using it.